packages/sdk/src/oracle/services/statemachines/switchOrAddChain.ts

```diff
@@ -0,0 +1,52 @@
+import { ethers } from "ethers";
+import Store from "../../store";
+import { Web3Provider } from "../../types/ethers";
+import { Handlers as GenericHandlers } from "../../types/statemachine";
+
+export type Params = {
+  chainId: number;
+  provider: Web3Provider;
+};
+
+export type Memory = undefined;
+
+export function initMemory(): Memory {
+  return undefined;
+}
+
+export function Handlers(store: Store): GenericHandlers<Params, Memory> {
+  return {
+    async start(params: Params) {
+      const { provider, chainId } = params;
+      try {
+        await provider.send("wallet_switchEthereumChain", [
+          {
+            chainId: ethers.utils.hexValue(chainId),
+          },
+        ]);
+        return "done";
+      } catch (err) {
+        if (err.code === -32603 || err.code === 4902) {
+          return "addAndSwitch";
+        }
+        throw err;
+      }
+    },
+    async addAndSwitch(params: Params) {
+      const { chainId, provider } = params;
+      const metadata = store.read().chainMetadata(chainId);
+      await provider.send("wallet_addEthereumChain", [
```

Comment:
What happens if the user rejects this add call? Does the next call fail gracefully? Should we detect that? Or does the error just show up in the console, so it's nothing to worry about?

Reply:
The nice thing about this statemachine architecture is that you can throw errors safely in these handlers, The higher level call will catch and attach it to the context.error property, which you can then inspect to see on the command what went wrong. This gives control to the front end dev to show appropriate errors to user when needed, or ignore completely.

heres a real example of the data in state that makes it really easy to debug what happened for each users action:
![image](https://user-images.githubusercontent.com/4429761/150995606-d8d76f4f-466a-4267-a330-b6312c747cfb.png)
The commands are numbeered in order and labeled by type, opening up the last approve request, we can see the user denied signing it.